File: packages/protocol-parser/src/encodeRecord.ts
import { StaticPrimitiveType, DynamicPrimitiveType } from "@latticexyz/schema-type";
import { Hex } from "viem";
import { encodeField } from "./encodeField";
import { Schema } from "./common";

/** @deprecated use `encodeValue` instead */
export function encodeRecord(
  valueSchema: Schema,
  values: readonly (StaticPrimitiveType | DynamicPrimitiveType)[]
): Hex {
  const staticValues = values.slice(0, valueSchema.staticFields.length) as readonly StaticPrimitiveType[];
  const dynamicValues = values.slice(valueSchema.staticFields.length) as readonly DynamicPrimitiveType[];

  const staticData = staticValues
    .map((value, i) => encodeField(valueSchema.staticFields[i], value).replace(/^0x/, ""))
    .join("");

  if (valueSchema.dynamicFields.length === 0) return `0x${staticData}`;

  const dynamicDataItems = dynamicValues.map((value, i) =>
    encodeField(valueSchema.dynamicFields[i], value).replace(/^0x/, "")
  );

  const dynamicFieldByteLengths = dynamicDataItems.map((value) => value.length / 2).reverse();
  const dynamicTotalByteLength = dynamicFieldByteLengths.reduce((total, length) => total + BigInt(length), 0n);

  const dynamicData = dynamicDataItems.join("");

  const packedCounter = `${dynamicFieldByteLengths
    .map((length) => encodeField("uint40", length).replace(/^0x/, ""))
    .join("")}${encodeField("uint56", dynamicTotalByteLength).replace(/^0x/, "")}`.padStart(64, "0");

  return `0x${staticData}${packedCounter}${dynamicData}`;
}

File: packages/services/pkg/protocol-parser/encodeRecord.go
package protocolparser

import (
	schematype "latticexyz/mud/packages/services/pkg/schema-type"
	"regexp"
	"strings"

	godash "github.com/golodash/godash/strings"
)

func EncodeRecord(schema Schema, values []interface{}) string {
	staticValues := values[0:len(schema.StaticFields)]
	dynamicValues := values[len(schema.StaticFields):]

	m1 := regexp.MustCompile(`^0x`)

	staticDataItems := []string{}
	for i, staticField := range schema.StaticFields {
		encoding := EncodeField(staticField, staticValues[i])
		encoding = m1.ReplaceAllString(encoding, "")
		staticDataItems = append(staticDataItems, encoding)
	}
	staticData := strings.Join(staticDataItems, "")

	if len(schema.DynamicFields) == 0 {
		return "0x" + staticData
	}

	dynamicDataItems := []string{}
	for i, dynamicField := range schema.DynamicFields {
		encoding := EncodeField(dynamicField, dynamicValues[i])
		encoding = m1.ReplaceAllString(encoding, "")
		dynamicDataItems = append(dynamicDataItems, encoding)
	}

	dynamicFieldByteLengths := []int{}
	for _, value := range dynamicDataItems {
		dynamicFieldByteLengths = append(dynamicFieldByteLengths, len(value)/2)
	}
	dynamicTotalByteLength := 0
	for _, length := range dynamicFieldByteLengths {
		dynamicTotalByteLength += length
	}

	dynamicData := strings.Join(dynamicDataItems, "")

	packedCounter := ""
	encodedDynamicTotalByteLength := EncodeField(schematype.UINT56, uint64(dynamicTotalByteLength))
	packedCounter += m1.ReplaceAllString(encodedDynamicTotalByteLength, "")

	for _, length := range dynamicFieldByteLengths {
		encodedLength := EncodeField(schematype.UINT40, uint64(length))
		packedCounter += m1.ReplaceAllString(encodedLength, "")
	}
	packedCounter = godash.PadEnd(packedCounter, 64, "0")

	return "0x" + staticData + packedCounter + dynamicData
}
